### ISSUE TITLE
Privately passing passphrases from scripts / shell hooks

### ISSUE THREAD

**Issue by jpouellet** [tedu/reop#19]
I dislike a916132 as I do not consider passing environment variables to be private.

Consider the following:
```
while true; do ps -eww; done | grep reop
```

I understand wanting to simplify things, but there is now no way to safely pass a passphrase unless you use the library interface directly, which is not always feasible if you're using reop from some shell-only hook in other software (you'd need some command line wrapper... and isn't that the point of reop(1)?)

**Comment by tedu** [comment]
Most operating systems should provide secure environments now; although I didn't do an exhaustive search. If there are known counterexamples, we can disable this functionality there.

I don't think pipes are any more secure. It's hard to protect a user (account) from itself. Every OS also provides ktrace/strace like facilities that could be used to sniff the password from stdin. (I tend to run ktrace much more often than ps -e; in terms of accidental exposure I personally would benefit from an environment setting, but that's hardly definitive.)

That said, I've considered an option to allow specifying a password file (which could be "-" or /dev/fd/X). The main objection is I didn't want to commit to another command line option. And what to call it? Perhaps -z?

Long term, some sort of reop-agent may be desirable, but I'd consider that more of a convenience than a security feature.

**Comment by jpouellet** [comment]
Fair enough. I am under the impression that historically systems have restricted whatever form of process tracing they possess more than they have prevented environment enumeration, but I suppose it's a weak argument anyway.

**Comment by tedu** [comment]
We can keep this open. It's worth thinking about.

**Comment by jpouellet** [comment]
Precedent (kind of): [encrypt(1)](http://www.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man1/encrypt.1) discourages passing passwords as an arg on the command line instead of via stdin. It is my understanding that visibility of environment variables is closer to visibility of arguments than that of data in pipes. Food for thought.

**Comment by jcs** [comment]
OpenBSD has only restricted viewing other user process environments since [5.3](http://cvsweb.openbsd.org/cgi-bin/cvsweb/src/sys/kern/kern_sysctl.c?rev=1.230&content-type=text/x-cvsweb-markup).  I always assumed storing secrets in the environment was insecure until I realized OpenBSD was the odd man out, but I still think it's a bad idea.

Environments get leaked in weird places (cron `X-Cron-Env` e-mail headers, web app exception handlers, etc.), although I guess it would be weird to be storing a reop password there.
